### ISSUE TITLE
Prevent ajax `adding_to_cart` conditionally

### ISSUE THREAD

**Issue by helgatheviking** [woocommerce/woocommerce#25759]
**Is your feature request related to a problem? Please describe.**

The ajax add to cart feature send an ajax request when the ajax add to cart button is clicked. I would like a way to prevent this ajax call from firing if certain conditions are not met. 

Here's the script [src](https://github.com/woocommerce/woocommerce/blob/master/assets/js/frontend/add-to-cart.js#L59-L107)

And there's a trigger in there just beforehand, but as far as I know, there's no way to hook into that trigger and prevent the code from proceeding.

```
// Trigger event.
$( document.body ).trigger( 'adding_to_cart', [ $thisbutton, data ] );
```

While generally simple products require no validation, there are times when plugins (like Name Your Price) might add fields that need validating, yet still want to take advantage of ajax add to cart.

In PHP we could use `apply_filters` to validate the product (and you do use that validating via `woocommerce_add_to_cart_validation`). I'd like to propose a JS alternative. 

**Describe the solution you'd like**
Some kind of condition/flag that will prevent the add to cart ajax call from firing. I'm proposing the same thing in One Page Checkout and it could make sense if you both used the same condition.

**Comment by juliaamosova** [comment]
Hi @helgatheviking,

Thank you for taking the time to report this bug, we really appreciate your help. Thanks for submitting a PR as well. I will leave it in the hands of the PR reviewer to provide further feedback.

**Comment by helgatheviking** [comment]
Sure thing! I think it will allow folks to use Name Your Price and ajax add to cart together! But hopefully someone more knowledgeable can advise me about replacing `trigger` with `triggerHandler` or if I need to have a new `triggerHandler` in _addition_ to the existing `trigger`.

**Comment by helgatheviking** [comment]
So I realized that I changed `trigger` to `triggerHandler` to make this work. Not sure if that's backcompatible, but on `trigger` it doesn't seem to accept returning a boolean value and so it wouldn't do what I needed.

As an alternative, we could leave the `trigger` in place and add a new `triggerHandler` like so:
```
// Trigger event.
$( document.body ).trigger( 'adding_to_cart', [ $thisbutton, data ] );

var valid = $( document.body ).triggerHandler( 'validate_adding_to_cart', [ $thisbutton, data ] );

if( valid !== false ) {
        // Do ajax
}
```

Let me know if you prefer I add a new commit for this, or submit a clean PR.
